### ISSUE TITLE
Issue on git with SSH and async

### ISSUE THREAD

**Issue by eloycoto** [ansible/ansible-modules-core#2625]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

git module
##### ANSIBLE VERSION

1.9
##### SUMMARY

We are heavy users of Ansible. But today I found a problem that I don't know how to fix it. [We are running a lot of task using async](http://acalustra.com/acelerate-your-ansible-playbooks-with-async-tasks.html), but with GIT we have an issue when we used with ssh instead of http. 

This is a small example of https git use: 

**Playbook:**

```
  - name: GIT CLONE
    git:
      repo: "https://github.com/kamailio/kamailio.git"
      dest: "/tmp/kamailio"
    async: 1000
    poll: 0
    register: loo_sleeper

  - name: "GIT CLONE CHECK THAT WAS OK"
    async_status: jid={{ loo_sleeper.ansible_job_id }}
    register: job_result
    until: job_result.finished
    retries: 300
```

**Https result:** 

```
TASK [GIT CLONE] ***************************************************************
ok: [127.0.0.1]

TASK [GIT CLONE CHECK THAT WAS OK] *********************************************
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (299 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}

FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (298 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (297 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (296 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (295 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (294 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (293 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (292 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (291 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/332426597965.1426', u'ansible_job_id': u'332426597965.1426', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'332426597965.1426'}}, u'changed': False}
changed: [127.0.0.1]
```

**SSH playbook**

```
  - name: GIT CLONE
    git:
      repo: "git@github.com:eloycoto/statsd-influxdb-examples.git"
      dest: "/tmp/test"
      key_file: "{{ project['root'] }}/keys/id_rsa"
      accept_hostkey: "yes"
    async: 100
    poll: 0
    register: loo_sleeper

  - name: "GIT CLONE CHECK THAT WAS OK"
    async_status: jid={{ loo_sleeper.ansible_job_id }}
    register: job_result
    until: job_result.finished
    retries: 300
```

**SSH result**

```
    TASK [GIT CLONE] ***************************************************************
    ok: [127.0.0.1]

    TASK [GIT CLONE CHECK THAT WAS OK] *********************************************
    FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (299 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/454442748921.28520', u'ansible_job_id': u'454442748921.28520', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'454442748921.28520'}}, u'changed': False}
    FAILED - RETRYING: TASK: GIT CLONE CHECK THAT WAS OK (298 retries left). Result was: {u'started': 1, u'finished': 0, u'results_file': u'/root/.ansible_async/454442748921.28520', u'ansible_job_id': u'454442748921.28520', 'invocation': {'module_name': u'async_status', 'module_args': {u'jid': u'454442748921.28520'}}, u'changed': False}
    fatal: [127.0.0.1]: FAILED! => {"ansible_job_id": "454442748921.28520", "changed": false, "cmd": "/usr/bin/git ls-remote '' -t refs/tags/HEAD", "failed": true, "finished": 1, "msg": "error: cannot run /home/vagrant/.ansible/tmp/ansible-tmp-1449220738.29-120669049111985/pT7ZsR: No such file or directory\nfatal: unable to fork", "rc": 128, "stderr": "error: cannot run /home/vagrant/.ansible/tmp/ansible-tmp-1449220738.29-120669049111985/pT7ZsR: No such file or directory\nfatal: unable to fork\n", "stdout": "", "stdout_lines": []}
```

If I use ssh without async works ok, but we don't want to block the deploy. It takes few minutes to download the repo and this is why we want async. 

So does anyone know the solution to this problem? I can't find the problem in the code. 

I tried with ansible 1.9 and with ansible devel version. 

Regards.

**Comment by vmindru** [comment]
this not a bug not, or not yet. Can you try to use a key without a password?

**Comment by vmindru** [comment]
---
- hosts: local
  vars:
  tasks:
      - name: GIT CLONE
        git:
          repo: git@github.com:eloycoto/statsd-influxdb-examples.git
          dest: "/tmp/kamailio"
          key_file: "/home/vmindru/.ssh/gittest"
        async: 100 
        poll: 10
        register: loo_sleeper
    
#      - name: "GIT CLONE CHECK THAT WAS OK"
#        async_status: jid={{ loo_sleeper.ansible_job_id }}
#        register: job_result
#        until: job_result.finished
#        retries: 300


PLAY [local] ****************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [GIT CLONE] ************************************************************* 
failed: [localhost] => {"ansible_job_id": "506387799772.7347", "changed": false, "cmd": "/usr/bin/git ls-remote '' -t refs/tags/HEAD", "failed": true, "finished": 1, "rc": 128}
stderr: error: cannot run /home/vmindru/.ansible/tmp/ansible-tmp-1449434669.79-148220901765541/7Lw2Fa: No such file or directory
fatal: unable to fork

msg: error: cannot run /home/vmindru/.ansible/tmp/ansible-tmp-1449434669.79-148220901765541/7Lw2Fa: No such file or directory
fatal: unable to fork
<job 506387799772.7347> FAILED on localhost

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/vmindru/2625_ssh.retry

localhost                  : ok=2    changed=0    unreachable=0    failed=1

**Comment by vmindru** [comment]
looks like during async for some reason remote is not being passed 
```
408 def is_remote_tag(git_path, module, dest, remote, version):
409     cmd = '%s ls-remote %s -t refs/tags/%s' % (git_path, remote, version)
410     (rc, out, err) = module.run_command(cmd, check_rc=True, cwd=dest)
411     if version in out:
412         return True
413     else:
414         return False
415 
```
this results in 

```/usr/bin/git ls-remote '' -t refs/tags/HEAD```

insted of

```/usr/bin/git ls-remote 'git@github.com:eloycoto/statsd-influxdb-examples.git' -t refs/tags/HEAD```

VM

**Comment by vmindru** [comment]
for some reason the async module remove teh TMP dir and then epxects it to be there. 

```
diff --git a/lib/ansible/plugins/action/async.py b/lib/ansible/plugins/action/async.py
index 8a7175a..66a6df6 100644
--- a/lib/ansible/plugins/action/async.py
+++ b/lib/ansible/plugins/action/async.py
@@ -69,9 +69,11 @@ class ActionModule(ActionBase):
         async_cmd = " ".join([str(x) for x in [env_string, async_module_path, async_jid, async_limit, remote_module_path, argsfile]])
         result.update(self._low_level_execute_command(cmd=async_cmd))

+
         # clean up after
-        if tmp and "tmp" in tmp and not C.DEFAULT_KEEP_REMOTE_FILES:
-            self._remove_tmp_path(tmp)
+        #this cleanup occures to early ... removing just to test
+        #if tmp and "tmp" in tmp and not C.DEFAULT_KEEP_REMOTE_FILES:
+        #    self._remove_tmp_path(tmp)

         result['changed'] = True


```

**Comment by vmindru** [comment]
Seems that this remove job is trigered to early , this would need a deeper thinking on how to properly fix for now this might be a lowsy workaround. 

https://github.com/vmindru/ansible/commit/4687a1d8407e99249494430fc71fb53c59113d1e 

```
diff --git a/lib/ansible/modules/core b/lib/ansible/modules/core
index 1913476..15dcabe 160000
--- a/lib/ansible/modules/core
+++ b/lib/ansible/modules/core
@@ -1 +1 @@
-Subproject commit 191347676eea08817da3fb237f24cdbf2d16e307
+Subproject commit 15dcabe01231d43788c05152c80eaec41bcb6dad-dirty
diff --git a/lib/ansible/plugins/action/async.py b/lib/ansible/plugins/action/async.py
index 8a7175a..831ea17 100644
--- a/lib/ansible/plugins/action/async.py
+++ b/lib/ansible/plugins/action/async.py
@@ -19,6 +19,7 @@ __metaclass__ = type

 import json
 import random
+import time

 from ansible import constants as C
 from ansible.plugins.action import ActionBase
@@ -70,9 +71,9 @@ class ActionModule(ActionBase):
         result.update(self._low_level_execute_command(cmd=async_cmd))

         # clean up after
+        time.sleep(1)
         if tmp and "tmp" in tmp and not C.DEFAULT_KEEP_REMOTE_FILES:
             self._remove_tmp_path(tmp)

         result['changed'] = True
-
         return result

```

**Comment by vmindru** [comment]
@core what is the core team mailing list? I think this one needs some discussion  as what this has to be fixed, i think removing the temp folder on remote machine without knowing what is teh result of the task is a bit against async module, particulary for the "fire and forget" the one late is expecting to find the temp task id present on the system or maybe i am understanding somethign wrong. shouldnt we better move this remove job to the async_wraper that would remove it self after it's complete running?

**Comment by jimi-c** [comment]
Hi @vmindru, have you tried this on devel/2.0? The way async works there was changed a bit and may have resolved this issue.

**Comment by jctanner** [comment]
@vmindru all of the public mailing lists are here http://docs.ansible.com/ansible/community.html#mailing-list-information

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jctanner** [comment]
@vmindru any response to @jimi-c's question? 

needs_info

**Comment by vmindru** [comment]
@jimi-c i did not i , i will attempt to do this tomorrow.

**Comment by vmindru** [comment]
tested this on branch stable-2.1 , works fine.

**Comment by vmindru** [comment]
tested this on devel branch , works fine.

**Comment by vmindru** [comment]
updated to latest commits in stable-2.0 , tested, works fine.

**Comment by vmindru** [comment]
i was not able to test this on 1.9 cause the env-setup would somehow fail even if i have the main repo on stable-1.9 and the submodules on latest stable-1.9. 

tested on the packaged 2.1 works fine. I think this can be closed.

**Comment by jctanner** [comment]
I am unable to reproduce this with 1.9.6 or devel, so closing. If anyone is able to reproduce with a recent version of ansible, please file a new issue.
